### ISSUE TITLE
Running on Windows 10

### ISSUE THREAD

**Issue by wbadry** [google-research/football#214]
Hello,
Is it possible to run football on Windows 10? 

I would appreciate if steps could be added. Thanks

**Comment by vi3itor** [comment]
Hi! 
A couple of months ago I applied most of the changes to the original [third-party project](https://github.com/BazkieBumpercar/GameplayFootball), compiled it with `MSVC`, and ran on Windows. You can do it yourself if you check out [windows branch](https://github.com/vi3itor/GameplayFootball/tree/windows) in my repo.
Next, I compiled this project on Windows (see `windows` branch in my fork), but didn't have enough time to debug run-time errors. I hope that I'll spare some free time in the next couple of weeks to continue the work. But any help and PRs are always welcome.

**Comment by WesleyKhoiVo** [comment]
[build.txt](https://github.com/google-research/football/files/5365438/build.txt)

Hi @vi3itor, can you check my log build on Windows? I tried to build like your instruction but not successful.

**Comment by WesleyKhoiVo** [comment]
[build.txt](https://github.com/google-research/football/files/5365464/build.txt)
Hi @vi3itor, Can you help me to check my log build? I tried to build like your instruction but it not successful. Thanks you in advance

**Comment by WesleyKhoiVo** [comment]
Hi @vi3itor, I built successfully, thanks a lot

**Comment by qstanczyk** [comment]
Guys, would you be able to make the changes to the main branch and send pull request (so that others can benefit from Windows setup too)? Thanks.

**Comment by vi3itor** [comment]
I'll try to do it once I finish the paper.

**Comment by Pierrolo** [comment]
Hi, @vi3itor I followed your instruction for windows build on your repo, when I launch it nothing happens, a windows seems to appear and disappear right away... Anybody got that as well ?
Thx a lot

**Comment by Pierrolo** [comment]
Hi, I've tried to build it on Windows using @vi3itor windows branch, following the instructions. 
But when I ran `cmake --build . --parallel --config Release` nothing got created in the Release folder, but it got created in the Debug folder. And in this Debug folder, running `gameplayfootball.exe` doesn't do anything...  Even after running `cmake --build . --parallel --config Debug`. A new window appears and disapears immediatly.
Is there somehting I'm missing ? Is there a specific way to run it somehow ?

**Comment by WesleyKhoiVo** [comment]
Hi Pierrolo,
Without debug crash log, we can't help you further. This version CMake is just generated [Visual Studio Makefile projects](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html) (not related with new CMake from Visual Studio [2017](https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-150) and [2019](https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-160)). So you need to [add breakpoint](https://stackoverflow.com/questions/16969788/cmake-generated-vs-project-cant-break-point) to Visual Studio projects `.vcpxoj` to get more details about this crash.
But, with your information about this crash, I am thinking about this crash related with OpenGL 3.2 (information from this [site](https://www.indiedb.com/games/gameplay-football)) or recheck graphic card driver.

**Comment by Pierrolo** [comment]
Hi, Got it to work, thanks. For anybody out there with the same problem and the same level of command line and installation as me (which is pretty much 0) here is the key things I missed : 

- In VS19 don't forget the "Desktop C++ dev tools" when you install VS19
- update graphic cards drivers and OpenGL 3.27

Thanks a lot @WesleyKhoiVo and @vi3itor  :D

**Comment by vi3itor** [comment]
@Pierrolo Great that you figured it out.
I know that I promised some time ago to finish porting to Windows, but it looks like soon I'll finally have time to do it.

**Comment by Pierrolo** [comment]
Whenever you do, I'll be the first in line to try it ;)
And for my next trick I will find a way to use the python API start and run games....
Has anybody ever managed to do this ? Is there a simple quick fix ? Like taking @vi3itor built engine and copy paste it somewhere by any chance ?

**Comment by vi3itor** [comment]
No, the original engine was significantly modified to cut away unnecessary multithreading, but also to speed up the game by throwing away menus, audio, etc. There were made many optimizations too.
So @Pierrolo, if you copy-paste the compiled original engine, it won't work.
As I said before, I was able to compile the engine from this repository on Windows (see the following [commit](https://github.com/vi3itor/football/commit/7261b45f9824a1b410c31eb7df7e75c97f0956b3)), but I had some run-time errors when I tried to use generated `dll` (you have to rename it too, btw).
I'm afraid there is no quick fix but diligent work with the debugger (unless you have a lot of experience and can instantly spot the problem).

**Comment by vi3itor** [comment]
Hi @WesleyKhoiVo, @Pierrolo, @wbadry! 
Could you please test [my PR](https://github.com/google-research/football/pull/277) on your machine? You can first define variable `set DOWNLOAD_ENGINE=1` to test the engine without building (it will be similar to installing as a wheel). And next you can follow [the guide](https://github.com/vi3itor/football/blob/windows/gfootball/doc/compile_engine.md#windows) and try compiling it. 
If you have any problems, let me know.
Thank you!

**Comment by WesleyKhoiVo** [comment]
Hi @vi3itor, I have a quick review but not build this PR (I don't have time right now, but will send you result when I have a time)

**Comment by WesleyKhoiVo** [comment]
I built your PR and ran successfully (Google Football). If I want to optimize performance, should I take source Google Football or Gameplay Football?

**Comment by WesleyKhoiVo** [comment]
I built your PR and ran successfully (Google Football)

**Comment by WesleyKhoiVo** [comment]
I built your PR and ran successfully (Google Football repo)

**Comment by xasans** [comment]
Hi @WesleyKhoiVo 
Do you compile exactly Google Research version(with python) or simple version( by documentation of @vi3itor )?
If you have compiled Google Research version, can you share with quick guides?

**Comment by vi3itor** [comment]
@xasans 
Do you have any troubles with compiling and installing `gfootball` by following the [guide](https://github.com/vi3itor/football/blob/windows/gfootball/doc/compile_engine.md#windows)? These changes (with minor improvements) will be merged into `master` branch once I finish with publishing the wheels.

**Comment by WesleyKhoiVo** [comment]
Hi @xasans, I just did what Vi3itor's instruction [here](https://github.com/vi3itor/football/blob/windows/gfootball/doc/compile_engine.md#windows). Remember your computer need meet hardware requirement and install compiler and library was listed by Vi3itor.

**Comment by xasans** [comment]
Thanks @vi3itor @WesleyKhoiVo

**Comment by xasans** [comment]
Hi, @vi3itor I have tried compiling by [your](https://github.com/vi3itor/football/blob/windows/gfootball/doc/compile_engine.md#windows) instruction but getting error after "Run compilation and installation"  command.


 Building wheel for gfootball (setup.py) 
  ERROR: Command errored out with exit status 1:
   command: 'c:\dev\vcpkg\football-env\scripts\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\dev\\vcpkg\\football\\setup.py'"'"'; __file__='"'"'C:\\dev\\vcpkg\\football\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\User\AppData\Local\Temp\pip-wheel-772i8pad'
       cwd: C:\dev\vcpkg\football\
  Complete output (114 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.9
  creating build\lib.win-amd64-3.9\gfootball
  copying gfootball\dump_to_txt.py -> build\lib.win-amd64-3.9\gfootball
  copying gfootball\dump_to_video.py -> build\lib.win-amd64-3.9\gfootball
  copying gfootball\play_game.py -> build\lib.win-amd64-3.9\gfootball
  copying gfootball\replay.py -> build\lib.win-amd64-3.9\gfootball
  copying gfootball\__init__.py -> build\lib.win-amd64-3.9\gfootball
  creating build\lib.win-amd64-3.9\gfootball\env
  copying gfootball\env\config.py -> build\lib.win-amd64-3.9\gfootball\env
  copying gfootball\env\constants.py -> build\lib.win-amd64-3.9\gfootball\env
  copying gfootball\env\controller_base.py -> build\lib.win-amd64-3.9\gfootball\env 
  ,,,

**Comment by vi3itor** [comment]
Can you please run 
`python -m pip install . --use-feature=in-tree-build --log log.txt`
and attach`log.txt` file here?

**Comment by xasans** [comment]
Here
[log.txt](https://github.com/google-research/football/files/6869636/log.txt)

**Comment by WesleyKhoiVo** [comment]
Can you attach `C:\Users\User\AppData\Local\Temp\pip-record-60rcle1b\install-record.txt` file for more detail?

**Comment by xasans** [comment]
Hi @WesleyKhoiVo  I can't find 'pip-record-60rcle1b' folder

**Comment by xasans** [comment]
What is hardware requirement for compiling?

**Comment by WesleyKhoiVo** [comment]
So weird, I found that link folder from your log.txt. I hoped that file contain show useful information, now we must find other clue.

Can you check `C:\Users\User\AppData\Local\Programs\Python\Python39\Include\gfootball` has a file name relate something like "log"?

Did you ever build a plain game? I mean game without python, see [vi3itor branch](https://github.com/vi3itor/GameplayFootball/tree/windows)

I think hardware requirement is depend on Visual Studio, which must be at least [2013](https://github.com/microsoft/vcpkg/pull/1752) for Vcpkg and then Visual Studio 2013 [hardware requirement here](https://stackoverflow.com/a/23742559/8707331).

**Comment by xasans** [comment]
Unfortunately  'gfootball' folder isn't exist.
Yes i already build game without python on this pc a month ago

**Comment by vi3itor** [comment]
@xasans Sorry for the confusion. When I drafted a PR I have changed the path in the instructions to the original repository. But since it had not been merged yet, you were trying to compile old version (2.9) which doesn't have Windows support yet. But it can be easily solved by replacing `git clone https://github.com/google-research/football.git` in the instructions with any of the options below. 
1. Either fetch and checkout my pull request in `google-research` repo:
```bash
git clone https://github.com/vi3itor/football.git
cd football
git checkout windows
```
2. Or checkout `windows` branch in my fork:
```bash
git clone https://github.com/google-research/football.git
cd football
git fetch origin pull/277/head:windows-pr
git checkout windows-pr
```
